atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/monitoring/TimestampTracker.java

```diff
@@ -0,0 +1,96 @@
+import com.codahale.metrics.Gauge;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.cleaner.Cleaner;
+import com.palantir.atlasdb.util.MetricsManager;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.logsafe.SafeArg;
+
+public class TimestampTracker implements AutoCloseable {
+    private static final Logger log = LoggerFactory.getLogger(TimestampTracker.class);
+
+    // We cache underlying calls, in case a hyper-aggressive metrics client repeatedly queries the values.
+    @VisibleForTesting
+    static final Duration CACHE_INTERVAL = Duration.ofSeconds(10L);
+
+    private final MetricsManager metricsManager = new MetricsManager();
+    private final Set<String> registeredMetricShortNames = Sets.newConcurrentHashSet();
+
+    private final Clock clock;
+
+    public TimestampTracker(Clock clock) {
+        this.clock = clock;
+    }
+
+    public static TimestampTracker createWithDefaultTrackers(TimelockService timeLockService, Cleaner cleaner) {
+        TimestampTracker tracker = new TimestampTracker(Clock.defaultClock());
+        tracker.registerTimestampForTracking("timestamp.fresh", timeLockService::getFreshTimestamp);
+        tracker.registerTimestampForTracking("timestamp.immutable", timeLockService::getImmutableTimestamp);
+        tracker.registerTimestampForTracking("timestamp.unreadable", cleaner::getUnreadableTimestamp);
+        return tracker;
+    }
+
+    @VisibleForTesting
+    void registerTimestampForTracking(String shortName, Supplier<Long> supplier) {
+        if (!putMetricIfAbsent(shortName)) {
+            log.warn("A metric with the name {} has already been registered in the timestamp tracker",
+                    SafeArg.of("metricName", shortName));
+            throw new IllegalStateException("A metric with the name " + shortName + " has already been registered!");
+        }
+        metricsManager.registerMetric(TimestampTracker.class, shortName, createCachingGauge(supplier));
+    }
+
+    private boolean putMetricIfAbsent(String shortName) {
+        synchronized (registeredMetricShortNames) {
+            if (registeredMetricShortNames.contains(shortName)) {
+                return false;
+            }
+            registeredMetricShortNames.add(shortName);
+            return true;
+        }
+    }
+
+    private <T> Gauge<T> createCachingGauge(Supplier<T> supplier) {
+        return new CachedGauge<T>(clock, CACHE_INTERVAL.getSeconds(), TimeUnit.SECONDS) {
+            @Override
+            protected T loadValue() {
+                return supplier.get();
+            }
+        };
+    }
+
+    @Override
+    public void close() {
+        // It is usually assumed that AtlasDB clients only have one TransactionManager open at a time,
+        // so deregistering everything starting with TimestampTracker is probably safe.
+        metricsManager.deregisterMetricsWithPrefix(TimestampTracker.class, "");
```

Comment:
we don't really support multiple transaction managers, so yeah this seems fine. we would end up registering metrics twice in that case (multiple transaction managers), right?

Reply:
I think internal auth product might do that / used to have done that when doing migrations actually (they've noticed some issues in the past e.g. #2099 that really only become significant if you do that). 

I guess they might get around this by only setting the Atlas registry to actually match theirs  later on?